### PR TITLE
Improve Charts compatibility with older and upcoming VTK versions.

### LINKED ISSUE #2841: Fix Charts test for VTK 9.2.0
https://github.com/pyvista/pyvista/pull/2837 caught that the VTK 9.2.0 release candidate will break, among others, a charts test (cc @dcbr).

The test breaks here:
https://github.com/pyvista/pyvista/blob/36ab8b59328f08dd76cf8a0f75db0b42555d4461/tests/test_charts.py#L299-L302

As I understand this part of the test checks that `axis.tick_locations = None` restores the "atomatically determined" tick locations. This is the error log:
```
==================================================================================================== FAILURES =====================================================================================================
____________________________________________________________________________________________________ test_axis ____________________________________________________________________________________________________

chart_2d = <Chart2D(0x4661bc0) at 0x7f7613fa2760>

    @skip_no_plotting
    def test_axis(chart_2d):
        l = "Y axis"
        r_fix, r_auto = [2, 5], None
        m = 50
        tc = 10
        tlabels = ["Foo", "Blub", "Spam"]
        tlocs, tlocs_large = [1, 5.5, 8], [5.2, 340, 9999.999]
        ts = 5
        tlo = 10
    
        # Test constructor arguments
        axis = charts.Axis(label=l, range=r_fix, grid=True)
        assert axis.label == l
        assert np.allclose(axis.range, r_fix) and axis.behavior == "fixed"
        assert axis.grid
    
        # Test properties, using the y axis of a 2D chart
        chart_2d.line([0, 1], [1, 10])
        chart_2d.show()
        axis = chart_2d.y_axis
    
        axis.label = l
        assert axis.label == l
        assert axis.GetTitle() == l
    
        axis.label_visible = False
        assert not axis.label_visible
        assert not axis.GetTitleVisible()
    
        axis.range = r_auto
        assert axis.behavior == "auto"
        axis.range = r_fix
        r = [0.0, 0.0]
        axis.GetRange(r)
        assert np.allclose(axis.range, r_fix)
        assert np.allclose(r, r_fix)
        assert axis.behavior == "fixed"
    
        assert axis.GetBehavior() == charts.Axis.BEHAVIORS["fixed"]
        axis.behavior = "auto"
        assert axis.behavior == "auto"
        assert axis.GetBehavior() == charts.Axis.BEHAVIORS["auto"]
        with pytest.raises(ValueError):
            axis.behavior = "invalid"
    
        axis.margin = m
        assert axis.margin == m
        assert axis.GetMargins()[0] == m
    
        axis.log_scale = True  # Log scale can be enabled for the currently drawn plot
        chart_2d.show()  # We have to call show to update all chart properties (calls Update and Paint methods of chart/plot objects).
        assert axis.log_scale
        assert axis.GetLogScaleActive()
        axis.log_scale = False
        chart_2d.show()
        assert not axis.log_scale
        assert not axis.GetLogScaleActive()
        # TODO: following lines cause "vtkMath::Jacobi: Error extracting eigenfunctions" warning to be printed.
        #  This is a VTK issue that will be fixed once PR (!8618) is merged.
        chart_2d.line([0, 1], [-10, 10])  # Plot for which log scale cannot be enabled
        axis.log_scale = True
        chart_2d.show()
        assert not axis.log_scale
        assert not axis.GetLogScaleActive()
    
        axis.grid = False
        assert not axis.grid
        assert not axis.GetGridVisible()
    
        axis.visible = False
        assert not axis.visible
        assert not axis.GetAxisVisible()
        axis.toggle()
        assert axis.visible
        assert axis.GetAxisVisible()
    
        tc0 = axis.tick_count
        axis.tick_count = tc
        assert axis.tick_count == tc
        assert axis.GetNumberOfTicks() == tc
        axis.tick_count = None
        assert axis.tick_count == tc0
        assert axis.GetNumberOfTicks() == tc0
        axis.tick_count = -1
        assert axis.tick_count == tc0
        assert axis.GetNumberOfTicks() == tc0
    
        tlocs0 = axis.tick_locations
        tlabels0 = axis.tick_labels
        axis.tick_locations = tlocs
        axis.tick_labels = tlabels
        assert np.allclose(axis.tick_locations, tlocs)
        assert np.allclose(axis.GetTickPositions(), tlocs)
        assert tuple(axis.tick_labels) == tuple(tlabels)
        assert vtk_array_to_tuple(axis.GetTickLabels()) == tuple(tlabels)
        axis.tick_labels = "2f"
        chart_2d.show()
        assert tuple(axis.tick_labels) == tuple(f"{loc:.2f}" for loc in tlocs)
        assert vtk_array_to_tuple(axis.GetTickLabels()) == tuple(f"{loc:.2f}" for loc in tlocs)
        assert axis.GetNotation() == charts.Axis.FIXED_NOTATION
        assert axis.GetPrecision() == 2
        axis.tick_labels = "4e"
        axis.tick_locations = tlocs_large  # Add some more variety to labels
        chart_2d.show()
        assert tuple(axis.tick_labels) == tuple(to_vtk_scientific(f"{loc:.4e}") for loc in tlocs_large)
        assert vtk_array_to_tuple(axis.GetTickLabels()) == tuple(
            to_vtk_scientific(f"{loc:.4e}") for loc in tlocs_large
        )
        assert axis.GetNotation() == charts.Axis.SCIENTIFIC_NOTATION
        assert axis.GetPrecision() == 4
        axis.tick_locations = None
        axis.tick_labels = None
        chart_2d.show()
>       assert np.allclose(axis.tick_locations, tlocs0)

tests/test_charts.py:302: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
<__array_function__ internals>:180: in allclose
    ???
../tmp_test_env_py39/lib/python3.9/site-packages/numpy/core/numeric.py:2251: in allclose
    res = all(isclose(a, b, rtol=rtol, atol=atol, equal_nan=equal_nan))
<__array_function__ internals>:180: in isclose
    ???
../tmp_test_env_py39/lib/python3.9/site-packages/numpy/core/numeric.py:2361: in isclose
    return within_tol(x, y, atol, rtol)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

x = array([-10.,  -8.,  -6.,  -4.,  -2.,   0.,   2.,   4.,   6.,   8.,  10.]), y = array([], dtype=float64), atol = 1e-08, rtol = 1e-05

    def within_tol(x, y, atol, rtol):
        with errstate(invalid='ignore'):
>           return less_equal(abs(x-y), atol + rtol * abs(y))
E           ValueError: operands could not be broadcast together with shapes (11,) (0,)

../tmp_test_env_py39/lib/python3.9/site-packages/numpy/core/numeric.py:2342: ValueError
```

Digging into this interactively, the problem is that we expect an empty tuple `()` to be set after `axis.tick_locations = None`, and instead we get `(-10.0, -9.0, -8.0, -7.0, -6.0, -5.0, -4.0, -3.0, -2.0, -1.0, 0.0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0)` (well, I got _this_ tuple from my interactive run that didn't set the window size to be the same as the test, which should explain the difference between this tuple and the values printed in the above traecback).

The interesting part is that temporarily disabling off-screen rendering in `tests/plotting/conftest.py` (which makes the chart figures pop up one after the other) makes the test pass! In that case the "expected default" captured earlier (`tlocs0`) in the test is _also_  `(-10.0, -9.0, -8.0, -7.0, -6.0, -5.0, -4.0, -3.0, -2.0, -1.0, 0.0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0)`.

The change with respect to VTK 9.1.0 is that in the older version we do get the empty tuple `()` both for `tlocs0` and `axis.tick_locations`, and both for on and off-screen rendering. There have been some fairly recent changes in matplotlib about computing tick positions (that I've run into), and I wonder if something in VTK was changed to accommodate that, and this is what breaks our test. (To be clear matplotlib's version was fixed at 3.5.2 for every single run in this issue.)

I'm not sure yet what changed, and more importantly, if we should fix the test or change something in the library. I did export the figure in question to a file with both on and off-screen rendering, and despite the failed test the two figures are the exact same. This might hint at that we only have to change the test.

### PR DESCRIPTION
**Pull request by dcbr**
### Overview

Improve Charts compatibility with older (<9.0) and upcoming (9.2) VTK versions.

### Details

- Support for the upstream VTK Charts changes introduced in [vtk!8618](https://gitlab.kitware.com/vtk/vtk/-/merge_requests/8618). Among other things, this allows users to customize the location and dimensions of Pie and Box Charts.
- Chart background falls back to `GetBackgroundBrush` in older VTK versions now (<9.0). This increases compatibility with older versions, although with slightly erroneous coloring (see [vtk#18238](https://gitlab.kitware.com/vtk/vtk/-/issues/18238)).

### Todo

- [x] Fix tests, including #2841.
- [x] ~Support custom background/border coloring of the active chart.~ --> Will do this in #3756.

### CODE REVIEW

**Comment by dcbr** [comment]
Regarding the tests I have a question: As this PR introduces different behaviour for different VTK versions, should there be separate tests for each version (and how would that look like) or just tests for the latest version/behaviour?

**Comment by adeak** [comment]
Thanks! Please also cross-check with #2863 for potential overlaps (I haven't looked at either yet).

**Comment by dcbr** [comment]
> Thanks! Please also cross-check with #2863 for potential overlaps (I haven't looked at either yet).

That one looked good (the axis tests part), I'll merge those changes in here once it is accepted.

**Comment by adeak** [comment]
> Regarding the tests I have a question: As this PR introduces different behaviour for different VTK versions, should there be separate tests for each version (and how would that look like) or just tests for the latest version/behaviour?

There's existing practice of skipping tests on older versions where functionality is not there, e.g.
https://github.com/pyvista/pyvista/blob/36ab8b59328f08dd76cf8a0f75db0b42555d4461/tests/plotting/test_plotting.py#L48
https://github.com/pyvista/pyvista/blob/36ab8b59328f08dd76cf8a0f75db0b42555d4461/tests/plotting/test_plotting.py#L64

So I think testing for newest is fine. As soon as there's a VTK 9.2.0 release... not sure what to do with the release candidate...

**Comment by dcbr** [comment]
Ok, I will proceed like that then 👍
Not sure what the failing "Labeler / triage" test is for? The other documentation and docstringcheck tests fail because they are using vtk<9.2 (they run fine when using the release candidate version).

**Comment by adeak** [comment]
I think the labeler task is about https://github.com/pyvista/pyvista/issues/2082, in any case the failure is due to a resource error. I'll just label your PR manually, hopefully the check will become green after rebase and further work.

**Comment by dcbr** [comment]
Ready for review. The documentation build should succeed if it uses VTK 9.2. The reason why the coverage is so low is unclear to me at the moment. All of these properties are tested in `test_charts`, but maybe there is a mismatch in the used VTK version here as well?

**Review by banesullivan — APPROVED** [review]
Looks good to me! Thank you for addressing #3224 

FYI, I merged in main to catch this up and re-run CI since there was a failure

**Comment by banesullivan** [comment]
@dcbr, the docs are failing from these changes:

```
/home/runner/work/pyvista/pyvista/pyvista/plotting/charts.py:docstring of pyvista.plotting.charts.ChartBox:18: WARNING: Exception occurred in plotting pyvista-ChartBox-8
 from /home/runner/work/pyvista/pyvista/doc/api/plotting/charts/_autosummary/pyvista.ChartBox.rst:
Traceback (most recent call last):
  File "/home/runner/work/pyvista/pyvista/pyvista/ext/plot_directive.py", line 328, in _run_code
    exec(code, ns)
looking for now-outdated files... none found
  File "<string>", line 6, in <module>
  File "/home/runner/work/pyvista/pyvista/pyvista/plotting/charts.py", line 3969, in size
    raise ValueError(
ValueError: Cannot set ChartBox geometry, it fills up the entire viewport by default. Upgrade to VTK v9.2 or newer.
/home/runner/work/pyvista/pyvista/pyvista/plotting/charts.py:docstring of pyvista.plotting.charts.ChartBox:18: WARNING: Exception occurred in plotting pyvista-ChartBox-11
 from /home/runner/work/pyvista/pyvista/doc/api/plotting/charts/_autosummary/pyvista.ChartBox.rst:
Traceback (most recent call last):
  File "/home/runner/work/pyvista/pyvista/pyvista/ext/plot_directive.py", line 328, in _run_code
    exec(code, ns)
  File "<string>", line 6, in <module>
  File "/home/runner/work/pyvista/pyvista/pyvista/plotting/charts.py", line 3969, in size
    raise ValueError(
ValueError: Cannot set ChartBox geometry, it fills up the entire viewport by default. Upgrade to VTK v9.2 or newer.
/home/runner/work/pyvista/pyvista/pyvista/plotting/charts.py:docstring of pyvista.plotting.charts.ChartPie:18: WARNING: Exception occurred in plotting pyvista-ChartPie-8
 from /home/runner/work/pyvista/pyvista/doc/api/plotting/charts/_autosummary/pyvista.ChartPie.rst:
Traceback (most recent call last):
  File "/home/runner/work/pyvista/pyvista/pyvista/ext/plot_directive.py", line 328, in _run_code
    exec(code, ns)
  File "<string>", line 6, in <module>
  File "/home/runner/work/pyvista/pyvista/pyvista/plotting/charts.py", line 4254, in size
    raise ValueError(
ValueError: Cannot set ChartPie geometry, it fills up the entire viewport by default. Upgrade to VTK v9.2 or newer.
/home/runner/work/pyvista/pyvista/pyvista/plotting/charts.py:docstring of pyvista.plotting.charts.ChartPie:18: WARNING: Exception occurred in plotting pyvista-ChartPie-11
 from /home/runner/work/pyvista/pyvista/doc/api/plotting/charts/_autosummary/pyvista.ChartPie.rst:
Traceback (most recent call last):
  File "/home/runner/work/pyvista/pyvista/pyvista/ext/plot_directive.py", line 328, in _run_code
    exec(code, ns)
  File "<string>", line 6, in <module>
  File "/home/runner/work/pyvista/pyvista/pyvista/plotting/charts.py", line 4254, in size
    raise ValueError(
ValueError: Cannot set ChartPie geometry, it fills up the entire viewport by default. Upgrade to VTK v9.2 or newer.
```

**Comment by banesullivan** [comment]
Please note that there are additional failures from #3782

File: tests/test_charts.py [hunk 1]
```diff
@@ -23,8 +23,8 @@
     not can_create_mpl_figure(), reason="Cannot create a figure using matplotlib"
 )
 
-# skip all tests if VTK<9.1.0
-if pyvista.vtk_version_info < (9, 1):
+# skip all tests if VTK<9.2.0
+if pyvista.vtk_version_info < (9, 2):
     pytestmark = pytest.mark.skip
```
[the remark this answers is not in the document]
Comment (dcbr) [reply tests/test_charts.py:28]: I didn't touch that except for increasing the minor VTK version, but I can use the decorator style instead.

**Comment by dcbr** [comment]
> @dcbr, the docs are failing from these changes

The docs fail because they are using VTK 9.1.0 (and they need VTK 9.2.0 to showcase all chart features). I believe we need a new osmesa VTK wheel [here](https://github.com/pyvista/pyvista-wheels) and update this line accordingly (probably best to do in a separate PR?):

https://github.com/pyvista/pyvista/blob/55c9e96fb0a0fe6abae783aa91d16ba19c2b59f6/.github/workflows/docs.yml#L49

Additionally, I believe we have all these coverage complaints because also there we are using VTK 9.1.0 instead of the latest version (coupled with `python==3.10`):

https://github.com/pyvista/pyvista/blob/55c9e96fb0a0fe6abae783aa91d16ba19c2b59f6/.github/workflows/testing-and-deployment.yml#L193

**Comment by akaszynski** [comment]
> probably best to do in a separate PR?

Yes. We'll need a new wheel. I'm also tempted to run (or try to) the docs in Python 3.11 since it should be somewhat faster.

File: tests/test_charts.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dcbr) [reply tests/test_charts.py:28]: And to answer your question: it does indeed skip the whole module, that's why coverage seems so low (all chart tests skipped as the used VTK version is lower than 9.2). Now I'm actually wondering why this module skip was put there in the first place, some tests for sure should be able to run on older VTK versions (that's the purpose of this PR as well).

[the remark this answers is not in the document]
Comment (dcbr) [reply tests/test_charts.py:28]: So the reason this was added was testing problems with some older VTK versions that were missing a certain module (see #2012). I'm going to just keep it there (with the increased version number) to keep things safe and simple, as it worked just fine before and works just fine on the newest VTK version.

**Comment by banesullivan** [comment]
> Yes. We'll need a new wheel.

I just built and published a wheel for 9.2.5 here: https://github.com/pyvista/pyvista-wheels/blob/main/vtk_osmesa-9.2.5-cp39-cp39-linux_x86_64.whl

```
pip install https://github.com/pyvista/pyvista-wheels/raw/main/vtk_osmesa-9.2.5-cp39-cp39-linux_x86_64.whl
```

This build is from the current (as of writing this comment) `release` branch in VTK which is ~9.2.5. My build configurations make this an easy process: https://github.com/banesullivan/vtk-cmake

```py
git clone git@gitlab.kitware.com:vtk/vtk.git && cd vtk
git checkout release
git submodule update --init --recursive
mkdir build && cd build
# copy the `*.cmake` files from my build configurations
cmake -GNinja -C osmesa.cmake ..
ninja
python setup.py bdist_wheel
```

> I'm also tempted to run (or try to) the docs in Python 3.11 since it should be somewhat faster.

I'll try to upload a 3.11 wheel as well

FYI, VTK is soon to start publishing OSMesa wheels as well now that https://gitlab.kitware.com/vtk/vtk/-/merge_requests/9820 is finished

**Comment by dcbr** [comment]
> I'm also tempted to run (or try to) the docs in Python 3.11 since it should be somewhat faster.

I just installed Python 3.11.1 locally (Windows 10) and doc building seems to complete successfully with VTK 9.2.5. I'm getting following warnings during the build (not sure if related to the python or VTK version):

```bash
WARNING: 'pypandoc' not available. Using Sphinx-Gallery to convert rst text blocks to markdown for .ipynb files.
```

<details><summary>Something related to XML reading/writing</summary>

```bash
2023-01-11 22:17:42.886 ( 104.864s) [                ]       vtkXMLWriter.cxx:649    ERR| vtkXMLPolyDataWriter (0000016F6353D180): Error opening output file "C:\Users\bramd\AppData\Local\Temp\tmp2muy2yyq.vtp"
2023-01-11 22:17:42.893 ( 104.871s) [                ]       vtkXMLWriter.cxx:652    ERR| vtkXMLPolyDataWriter (0000016F6353D180): Error code "Permission denied"
2023-01-11 22:17:42.893 ( 104.871s) [                ]       vtkExecutive.cxx:741    ERR| vtkCompositeDataPipeline (0000016F63AC2790): Algorithm vtkXMLPolyDataWriter (0000016F6353D180) returned failure for request: vtkInformation (0000016F638DD080)
  Debug: Off
  Modified Time: 2590781
  Reference Count: 1
  Registered Events: (none)
  Request: REQUEST_DATA
  FORWARD_DIRECTION: 0
  ALGORITHM_AFTER_FORWARD: 1
  FROM_OUTPUT_PORT: -1


2023-01-11 22:17:42.894 ( 104.873s) [                ]       vtkXMLReader.cxx:317    ERR| vtkXMLPolyDataReader (0000016F4B25C860): Error opening file C:\Users\bramd\AppData\Local\Temp\tmp2muy2yyq.vtp
2023-01-11 22:17:42.895 ( 104.874s) [                ]       vtkExecutive.cxx:741    ERR| vtkCompositeDataPipeline (0000016F63AC1FB0): Algorithm vtkXMLPolyDataReader (0000016F4B25C860) returned failure for request: vtkInformation (0000016F638DE660)
  Debug: Off
  Modified Time: 2590874
  Reference Count: 1
  Registered Events: (none)
  Request: REQUEST_INFORMATION
  FORWARD_DIRECTION: 0
  ALGORITHM_AFTER_FORWARD: 1


2023-01-11 22:17:42.899 ( 104.877s) [                ]       vtkXMLReader.cxx:317    ERR| vtkXMLPolyDataReader (0000016F4B25B660): Error opening file C:\Users\bramd\AppData\Local\Temp\tmp2muy2yyq.vtp
2023-01-11 22:17:42.899 ( 104.877s) [                ]       vtkExecutive.cxx:741    ERR| vtkCompositeDataPipeline (0000016F63AC2F70): Algorithm vtkXMLPolyDataReader (0000016F4B25B660) returned failure for request: vtkInformation (0000016F638DD390)
  Debug: Off
  Modified Time: 2591026
  Reference Count: 1
  Registered Events: (none)
  Request: REQUEST_INFORMATION
  FORWARD_DIRECTION: 0
  ALGORITHM_AFTER_FORWARD: 1


2023-01-11 22:17:42.901 ( 104.879s) [                ]       vtkXMLReader.cxx:317    ERR| vtkXMLPolyDataReader (0000016F4B25B660): Error opening file C:\Users\bramd\AppData\Local\Temp\tmp2muy2yyq.vtp
2023-01-11 22:17:42.901 ( 104.880s) [                ]       vtkExecutive.cxx:741    ERR| vtkCompositeDataPipeline (0000016F63AC2F70): Algorithm vtkXMLPolyDataReader (0000016F4B25B660) returned failure for request: vtkInformation (0000016F638DD390)
  Debug: Off
  Modified Time: 2591026
  Reference Count: 1
  Registered Events: (none)
  Request: REQUEST_INFORMATION
  FORWARD_DIRECTION: 0
  ALGORITHM_AFTER_FORWARD: 1


2023-01-11 22:17:42.902 ( 104.881s) [                ]vtkStreamingDemandDrive:819    ERR| vtkCompositeDataPipeline (0000016F63AC2F70): No update piece number has been set in the information for output port 0 on algorithm vtkXMLPolyDataReader (0000016F4B25B660).
```

</details>

```bash
P:\python\venv\pyvista311\Lib\site-packages\zmq\_future.py:679: RuntimeWarning: Proactor event loop does not implement add_reader family of methods required for zmq. Registering an additional selector thread for add_reader support via tornado. Use `asyncio.set_event_loop_policy(WindowsSelectorEventLoopPolicy())` to avoid this warning.
  self._get_loop()
```

```bash
Assertion failed: pfd.revents & POLLIN (C:\projects\libzmq\src\signaler.cpp:265)
```

Finally, this one occurs multiple times throughout the build:
```bash
0.00s - Debugger warning: It seems that frozen modules are being used, which may
0.00s - make the debugger miss breakpoints. Please pass -Xfrozen_modules=off
0.00s - to python to disable frozen modules.
0.00s - Note: Debugging will proceed. Set PYDEVD_DISABLE_FILE_VALIDATION=1 to disable this validation.
```

Other than that, everything looks good though.

**Comment by adeak** [comment]
> WARNING: 'pypandoc' not available. Using Sphinx-Gallery to convert rst text blocks to markdown for .ipynb files.

That one I know (from Python 3.9): it's what it says. Presumably if you install `pypandoc` in your environment this will go away. I never install it (we don't have it in our requirements files), and all this does is raise that warning. I never check the notebooks generated by the doc build, it might differ from what `pypandoc` would do, but anyway this doesn't break the build, you can ignore it.

**Comment by dcbr** [comment]
Ready for review. The remaining codecov complaints are all related to "compatibility code" for VTK<9 (which passed before #3822, but fails now that VTK 9.2 is used for coverage reports).

**Comment by adeak** [comment]
> The remaining codecov complaints are all related to "compatibility code" for VTK<9 (which passed before #3822, but fails now that VTK 9.2 is used for coverage reports).

Should we add no-cover pragmas? VTK < 9 is becoming pretty obscure.

**Comment by dcbr** [comment]
> Should we add no-cover pragmas? VTK < 9 is becoming pretty obscure.

Definitely, I kind of forgot those existed. Added in last commit!

**Review by banesullivan — APPROVED** [review]

**Pull request merged**